### PR TITLE
mode: refactor code structure for item purchase

### PR DESCRIPTION
**Pull request by adamqqqplay**
Move all item_purchase_*.lua to subdirectory to reduce maintenance costs and make it easier for players to edit.

Fix: #126

### CODE REVIEW

**Comment by adamqqqplay** [comment]
@JoshdanG @ArcherWayne 
Please take a look at item_purchase_generic.lua and give me some advice on how to call ItemPurchaseSystem directly. The current implementation may introduce some additional overhead.

File: hero/_template.lua [hunk 1]
```diff
@@ -47,12 +48,13 @@ local ItemsToBuy =
...
 --检查装备列表
-ItemPurchaseSystem.checkItemBuild(ItemsToBuy)
+ItemPurchaseSystem:CreateItemInformationTable(GetBot(), ItemsToBuy)
 
 function ItemPurchaseThink()
 	--购买辅助物品	对于辅助英雄保留这一行
 	ItemPurchaseSystem.BuySupportItem()
-	--购买信使		对于5号位保留这一行
 	--购买装备
 	ItemPurchaseSystem.ItemPurchase(ItemsToBuy)
```
Comment (JoshdanG) [inline hero/_template.lua:57]:
```suggestion
	ItemPurchaseSystem.ItemPurchaseExtend()
```

File: item_purchase_generic.lua [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+end
+
+local HeroInfoFile = "NOT IMPLEMENTED"
+
+if bot:IsHero() then
+	HeroInfoFile = require(GetScriptDirectory() .. "/hero/" .. string.gsub(GetBot():GetUnitName(), "npc_dota_hero_", ""));
+end
+
+-- TODO: Call ItemPurchaseSystem directly
+function ItemPurchaseThink()
```
Comment (JoshdanG) [inline item_purchase_generic.lua:15]: You should already be able to just call `ItemPurchaseSystem.BuySupportItem()` and 	`ItemPurchaseSystem:ItemPurchaseExtend()` directly.  `CreateItemInformationTable` adds the `itemInformationTable` to the bot itself, so it's always there when you use `GetBot()`.  

I think if you want to be _extremely_ direct you can do
```
local ItemPurchaseSystem = dofile(GetScriptDirectory() .. "/util/ItemPurchaseSystem")
local ItemPurchaseThink = ItemPurchaseSystem:ItemPurchaseExtend
```
You'd have to incorporate / call `BuySupportItem()` in `ItemPurchaseExtend()`, but that probably makes sense to do anyway.

**Review by ArcherWayne — APPROVED** [review]

**Review by JoshdanG — COMMENTED** [review]
Does this actually work? Does item_purchase_generic get loaded separately for each hero?

I think the individual hero files should only contain X.ItemsToBuy and everything else should be handled centrally.  If needed you could add a "X.shouldBuySupportItems=true" and other similar attributes, but I think currently any variance there is accidental.

Just calling the folder "hero" seems wrong.  What's the end goal in terms of folder structure? There will likely be multiple files for each hero.  Probably you want either hero/items/hero_name.lua or hero/hero_name/items.lua

**Comment by adamqqqplay** [comment]
> Does this actually work? Does item_purchase_generic get loaded separately for each hero?

Thanks for your detailed review. It does work in my local tests. item_purchase_generic will run a separate instance for each hero, and find the corresponding item build list in the subdirectory through the hero's name.


> I think the individual hero files should only contain X.ItemsToBuy and everything else should be handled centrally. If needed you could add a "X.shouldBuySupportItems=true" and other similar attributes, but I think currently any variance there is accidental.

What you said is right, I also want to achieve this, but the current item purchase function is a little complicated, and I haven't seen the logic between them clearly yet.

> Just calling the folder "hero" seems wrong. What's the end goal in terms of folder structure? There will likely be multiple files for each hero. Probably you want either hero/items/hero_name.lua or hero/hero_name/items.lua

I think this directory can include the hero's ability building list and item building list, so I temporarily named it "hero". Your opinion is very useful, I will reconsider how to modify it.

File: item_purchase_generic.lua [hunk 3]
```diff
@@ -0,0 +1,17 @@
+local bot = GetBot()
+
+if bot:IsInvulnerable() or bot:IsHero() == false or bot:IsIllusion()
+then
+	return
+end
+
+local HeroInfoFile = "NOT IMPLEMENTED"
+
+if bot:IsHero() then
```
[the remark this answers is not in the document]
Comment (adamqqqplay) [reply item_purchase_generic.lua:10]: Ok, maybe there needs to be a default item buy list for heroes.

File: item_purchase_generic.lua (hunk 2 above)
Comment (adamqqqplay) [reply item_purchase_generic.lua:15]: Thanks a lot! After your explanation I finally understand what the CreateItemInformationTable function actually does.

Regarding `BuySupportItem()`, I think I will use RoleUtility to judge whether the hero is a support.

**Comment by JoshdanG** [comment]
I created a pull request to show what I was suggesting about reducing the file contents. https://github.com/adamqqqplay/dota2ai/pull/130

There was one snag I overlooked -- ItemPurchaseSystem was unnecessarily using "self" and ItemsToBuy, so I cleaned this up as well.

I have marked it as draft because I only did the change for 10 heros and set hero selection to debug so only they would be used. If you want to use it I can clean it up to be a proper pull request.

**Comment by adamqqqplay** [comment]
> I created a pull request to show what I was suggesting about reducing the file contents. #130
> 
> There was one snag I overlooked -- ItemPurchaseSystem was unnecessarily using "self" and ItemsToBuy, so I cleaned this up as well.
> 
> I have marked it as draft because I only did the change for 10 heros and set hero selection to debug so only they would be used. If you want to use it I can clean it up to be a proper pull request.

@JoshdanG Thanks very much! I think your work is very great. If you have time, you can merge these two PRs and submit it to the master branch. If you haven't gotten around to it yet, I'm going to take the time to deal with that this weekend.

**Comment by adamqqqplay** [comment]
Finished at #133

**Pull request closed without merging**